### ISSUE TITLE
NPM Overrides not taken into account

### ISSUE THREAD

**Issue by stumueller** [aquasecurity/trivy#4705]
### Discussed in https://github.com/aquasecurity/trivy/discussions/4704

<div type='discussions-op-text'>

<sup>Originally posted by **stumueller** June 23, 2023</sup>
### Description

There is already an issue for this #3816 However the topic is closed and I can't comment on it.

the new vulnerability in semver <7.5.2 has been flagged by Trivy.
My package.json contains

"overrides": {
    "semver": "^7.5.3"
  },

after doing an npm install my package-lock.json still makes references to semver < 7.5.2 as dependancies.

However an npm audit comes back clean. And an npm list --all shows any semver package as being overridden with 7.5.3
checking npm_modules does not return any semver package apart from 7.5.3

### Desired Behavior

Trivy would take into account the overrides section of package.json

### Actual Behavior

Trivy still reports vulnerabilities

### Debug Output

```bash
2023-06-23T15:45:26.982+0100    DEBUG   Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-06-23T15:45:26.986+0100    DEBUG   cache dir:  /home/stumueller/.cache/trivy
2023-06-23T15:45:26.987+0100    INFO    Need to update DB
2023-06-23T15:45:26.987+0100    INFO    DB Repository: ghcr.io/aquasecurity/trivy-db
2023-06-23T15:45:26.987+0100    INFO    Downloading DB...
37.86 MiB / 37.86 MiB [----------------------------------------------------------------------] 100.00% 7.74 MiB p/s 5.1s2023-06-23T15:45:32.937+0100    DEBUG   Updating database metadata...
2023-06-23T15:45:32.937+0100    DEBUG   DB Schema: 2, UpdatedAt: 2023-06-23 12:10:15.757757128 +0000 UTC, NextUpdate: 2023-06-23 18:10:15.757756628 +0000 UTC, DownloadedAt: 2023-06-23 14:45:32.937556793 +0000 UTC
2023-06-23T15:45:32.937+0100    INFO    Vulnerability scanning is enabled
2023-06-23T15:45:32.937+0100    DEBUG   Vulnerability type:  [os library]
2023-06-23T15:45:32.937+0100    INFO    Secret scanning is enabled
2023-06-23T15:45:32.937+0100    INFO    If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-06-23T15:45:32.937+0100    INFO    Please see also https://aquasecurity.github.io/trivy/v0.42/docs/secret/scanning/#recommendation for faster secret detection
2023-06-23T15:45:32.939+0100    DEBUG   No secret config detected: trivy-secret.yaml
2023-06-23T15:45:32.940+0100    DEBUG   Image ID: sha256:5f210ecfbc411743861023aab969fd1961e70f982f32aad89ed04365ec3c3962
2023-06-23T15:45:32.940+0100    DEBUG   Diff IDs: [sha256:78a822fe2a2d2c84f3de4a403188c45f623017d6a4521d23047c9fbb0801794c sha256:fed856459b5efb5ca8ba0ea035f4b3b7cf4954cf35949b4027ae03ab69e22244 sha256:2ed92fd7e053235a103c303aef532cbecd90176c67656610d815e92754e914e7 sha256:34aad8aaa364b41dadc4737ddba82355939a6a027d1cfdf6fd96df14cf602928 sha256:081da77da241640da0c349c066e3438efb1ca1664324ecd4e832238d83a27c00 sha256:edd9e8a27d39c2629fab73076589dccf9147fa1daf6bb831c0a50114481ac1be sha256:2dea5e390911febb3a3f69671c84bb9bed327d5bb1669eaa95bd11f67cc89781]
2023-06-23T15:45:32.940+0100    DEBUG   Base Layers: [sha256:78a822fe2a2d2c84f3de4a403188c45f623017d6a4521d23047c9fbb0801794c sha256:fed856459b5efb5ca8ba0ea035f4b3b7cf4954cf35949b4027ae03ab69e22244 sha256:2ed92fd7e053235a103c303aef532cbecd90176c67656610d815e92754e914e7 sha256:34aad8aaa364b41dadc4737ddba82355939a6a027d1cfdf6fd96df14cf602928]
2023-06-23T15:45:32.946+0100    DEBUG   Missing image ID in cache: sha256:5f210ecfbc411743861023aab969fd1961e70f982f32aad89ed04365ec3c3962
2023-06-23T15:45:32.947+0100    DEBUG   Missing diff ID in cache: sha256:081da77da241640da0c349c066e3438efb1ca1664324ecd4e832238d83a27c00
2023-06-23T15:45:32.947+0100    DEBUG   Missing diff ID in cache: sha256:2dea5e390911febb3a3f69671c84bb9bed327d5bb1669eaa95bd11f67cc89781
2023-06-23T15:45:32.947+0100    DEBUG   Missing diff ID in cache: sha256:edd9e8a27d39c2629fab73076589dccf9147fa1daf6bb831c0a50114481ac1be
2023-06-23T15:45:38.033+0100    DEBUG   No secrets found in container image config
2023-06-23T15:45:38.036+0100    INFO    Detected OS: alpine
2023-06-23T15:45:38.036+0100    INFO    Detecting Alpine vulnerabilities...
2023-06-23T15:45:38.036+0100    DEBUG   alpine: os version: 3.18
2023-06-23T15:45:38.036+0100    DEBUG   alpine: package repository: 3.18
2023-06-23T15:45:38.036+0100    DEBUG   alpine: the number of packages: 17
2023-06-23T15:45:38.037+0100    INFO    Number of language-specific files: 1
2023-06-23T15:45:38.037+0100    INFO    Detecting node-pkg vulnerabilities...
2023-06-23T15:45:38.037+0100    DEBUG   Detecting library vulnerabilities, type: node-pkg, path:
2023-06-23T15:45:38.048+0100    DEBUG   Found an ignore file .trivyignore
2023-06-23T15:45:38.048+0100    DEBUG   These IDs will be ignored: []
2023-06-23T15:45:38.048+0100    DEBUG   Found an ignore file .trivyignore
2023-06-23T15:45:38.048+0100    DEBUG   These IDs will be ignored: []

web (alpine 3.18.2)

Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)

2023-06-23T15:45:38.048+0100    INFO    Table result includes only package filenames. Use '--format json' option to get the full path to the package file.

Node.js (node-pkg)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

┌───────────────────────┬────────────────┬──────────┬───────────────────┬───────────────┬───────────────────────────────────────────────────────────┐
│        Library        │ Vulnerability  │ Severity │ Installed Version │ Fixed Version │                           Title                           │
├───────────────────────┼────────────────┼──────────┼───────────────────┼───────────────┼───────────────────────────────────────────────────────────┤
│ semver (package.json) │ CVE-2022-25883 │ MEDIUM   │ 7.3.7             │ 7.5.2         │ semver vulnerable to Regular Expression Denial of Service │
│                       │                │          │                   │               │ https://avd.aquasec.com/nvd/cve-2022-25883                │
└───────────────────────┴────────────────┴──────────┴───────────────────┴───────────────┴───────────────────────────────────────────────────────────┘
```

### Operating System

Ubuntu 22.4

### Version

```bash
0.42.1
```

### Checklist

- [ ] Run `trivy --reset`
- [ ] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)</div>

**Comment by knqyf263** [comment]
Maintainers create an issue after confirming the issue

**Comment by Arjun-M-291297** [comment]
Made following changes to override the semver version to bypass the Trivy scanner to accept a fixed version (7.5.2).
 
 "overrides": {
    "semver": "^7.5.2"
  },
  "resolutions": {
    "semver": "^7.5.2"
  },

When I tried to run ‘npm i’ in my local environment and searched for all occurrences of semver in package-lock.json ;
Under the packages "semver": "^7.5.2" version got updated.

But Trivy scanner is unable to take the updated version and still throws error in the build.

(alpine 3.17.4)
==============================================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)


Node.js (node-pkg)
==================
Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

┌───────────────────────┬────────────────┬──────────┬───────────────────┬───────────────┬───────────────────────────────────────────────────────────┐
│        Library        │ Vulnerability  │ Severity │ Installed Version │ Fixed Version │                           Title                           │
├───────────────────────┼────────────────┼──────────┼───────────────────┼───────────────┼───────────────────────────────────────────────────────────┤
│ semver (package.json) │ CVE-2022-25883 │ MEDIUM   │ 7.5.1             │ 7.5.2         │ semver vulnerable to Regular Expression Denial of Service │
│                       │                │          │                   │               │ https://avd.aquasec.com/nvd/cve-2022-25883                │
└───────────────────────┴────────────────┴──────────┴───────────────────┴───────────────┴───────────────────────────────────────────────────────────┘
